Added Units / Variables Support

TheRealSyler opened the pull request:
much cleaner and better version of the previous pr, plus added variables.

TheRealSyler committed:
Added Units / Variables Support

and changed sassSchema completionItem.insertText from type string to SnippetString.

minkir014 commented:
@TheRealSyler Does this PR provide variables completion across files??

TheRealSyler commented:
@minkir014  no, it generates the variables from the current file when the completion gets triggered and doesn't store them.

minkir014 commented:
Why?? But this feature is more needed than variables completion in the same file.

TheRealSyler commented:
@minkir014  i would have to load other files that isn't easy to do at least the way i tried, i also couldn't find good documentation, if you know a way please tell me, since this is a nice to have thing i rather spend my time on other things.

minkir014 commented:
See this extension it's for SCSS. It has completion for variables and mixins.

TheRealSyler commented:
don't know if you mean the extension that i just looked at, the way they do it is they cache the file variables, which would be possible to implement.

minkir014 commented:
Sorry, I forgot to paste the link.
https://github.com/mrmlnc/vscode-scss
Yes, they do that with cache files.

robinbentley commented:
@therealsyler I have some tasks down for a sass project tomorrow. I’ll use this branch and get back to you 👍🏼

robinbentley commented on src/schemas/sassSchemaUnits.ts:
```diff
@@ -0,0 +1,127 @@
+  {
+    name: 's',
+    body: '$1s',
+    description: '($value) - Time unit (seconds)'
```
We can probably remove the `($value) - ` from each description, it's not really offering much and just increases the size of the info box.

robinbentley commented on src/schemas/sassSchemaUnits.ts:
```diff
@@ -0,0 +1,127 @@
+const sassSchemaUnits = [
+  {
+    name: 's',
+    body: '$1s',
```
I don't feel it's necessary to have the tab stop return to the value,  especially since the snippet includes the value already. I would expect to be able to hit return or tab right after the snippet had been selected and this currently just separates the value from the unit.

robinbentley commented on src/sassAutocomplete.ts:
```diff
@@ -153,37 +143,86 @@ export function getValues(cssSchema, currentWord: string): CompletionItem[] {
+    }
     if (value) {
       values = getValues(cssSchema, currentWord);
+      variables = getVariables(text);
```
I can see variables being added to the completions list however the completion prompt is not displaying. This could be a settings issue on my end. 

As mentioned in the previous PR comments though I don't really see a huge value in being able to complete variables within the same file. Personally the majority of projects I have worked on (and as can be seen in most libraries) have had a single file for configuring variables to be used across the other project files. The benefit that jumps to mind is in a large config file like bootstraps where variables are being used to define others. I appreciate this portion wasn't part of your initial scope.

robinbentley reviewed (changes requested):
Hey @TheRealSyler, sorry it's taken me so long to get around to this. Thank you for contributing! 

Couple of specific comments below but generally what's here is technically good. The only real issues I have are practical ones which are mostly based on my workflow. 

I **finally** managed to get some time on a sass project and I find that having the option to complete units is quite intrusive, whilst newer devs may find that having an explanation of each unit is useful all it meant for me is that I end up with multiple extra keystokes just to move on after typing a unit. either through hitting `esc` to close the prompt then `enter/tab` to move on or hitting `enter` to select the completion and then `tab` to get out of the snippet before being able to `tab/enter` to get on my way. 

If this was going to be added to the extension my thought is that it will be an opt in feature via a settings flag. 

Let me know if you have any questions. 🙂

TheRealSyler committed:
fixed some issues and added variable completion for all files.

TheRealSyler committed:
added check so that the imported variables dont return null.

TheRealSyler commented:
@robinbentley i think, i fixed all the issues, the reason that the variables didn't show up is that you probably  typed $ and the completion removed the $ sign (now the $ sign doesn't get removed), i also added variable completion for other files, the variable completions of the current file get stored in the workspaceStare would have been great if knew before that such a thing existed anyways, the readme should have a note that the completions for variables in other files only get added once the completion function ran once, i can add it if there are no more issues.

minkir014 commented:
Will the extension after the commit 52f9f0c be able to add completion for variables across files???

robinbentley commented:
@TheRealSyler Thanks, I'm at a wedding tomorrow but I try to pull down and have another play on Sunday.

TheRealSyler committed:
fixing the get variables path

TheRealSyler commented:
@minkir014  yes.

@robinbentley sorry for all the commits i didn't test enough before pushing, because i was sure that there where no issues, now everything should work as expected.

minkir014 commented:
Won't this PR be merged????

TheRealSyler commented:
@minkir014 in case you want to try the new version now: https://drive.google.com/open?id=1uDVQ-sJaWuz-V_YwY63XzAPYL6oBAFi3
i use the variable completion a lot, its really a great change, also i will probably improve the extension even more in the next few days, if you have some other ideas i might put them on my todo list, if i feel that they would make the extension better.

robinbentley commented:
@TheRealSyler As you're intending to actively work on the extension how would you feel about taking over development and publishing?

I'm quite happy to un-publish my version from the market place, archive this repo, and direct them to yours. I started this extension when VSCode first came out because I needed the support for the indented syntax day to day. As I no longer work with it and with my increasing other commitments I just don't have the time to maintain the project.

All I would request is that the periods I was involved remained on the license (i.e `Copyright (c) 2016-2019 Robin Bentley` and a credit on the README, at least whilst code I contributed is still in use.

Let me know your thoughts.

Edit: Just to be clear, I don't mean to be the sole person on the license. You would add yourself above my entry of course

TheRealSyler commented:
@robinbentley sounds good, you could just transfer the repo instead of archiving it.

robinbentley commented:
@TheRealSyler I'd like to retain it. Have you deleted your fork or is it just private?

TheRealSyler commented:
@robinbentley  i have deleted it for  now.

robinbentley commented:
@TheRealSyler Ok, cool. this evening I will merge to a branch, you're free to fork again and do with it as you please. Once you're set I'm going to un-publish from the VSCode marketplace and it's all yours. 

[https://code.visualstudio.com/api/working-with-extensions/publishing-extension](https://code.visualstudio.com/api/working-with-extensions/publishing-extension)

TheRealSyler commented:
@robinbentley Great , im ready when you are.

robinbentley commented:
@TheRealSyler The content of this PR can be found in this branch https://github.com/robinbentley/vscode-sass-indented/tree/pr63 

Re-fork the repo and let me know when you are ready to publish on your publisher account for I can remove it from mine.

TheRealSyler commented:
@robinbentley im ready.

robinbentley commented:
@TheRealSyler I'm unpublished, I'll update this repo's README now. Cheers for keeping it going, best of luck!

TheRealSyler commented:
@TheRealSyler i just published, https://marketplace.visualstudio.com/items?itemName=Syler.sass-indented
